File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []

        def dfs(i, cur, total):
            # base case of success
            if total == target:
                res.append(cur.copy())
                return
            # base case of OOB or past target
            if i >= len(nums) or total > target:
                return

            # recursive steps

            # inclusion
            cur.append(nums[i])
            dfs(i, cur, total + nums[i])
            cur.pop()

            # exclusion
            dfs(i + 1, cur, total)

        dfs(0,[],0)
        return res

File: Data Structures & Algorithms/maximum-subarray/submission-0.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if(nums.size() == 1) return nums[0];

        int res = nums[0];
        int maxEnding = nums[0];

        for(int i = 1; i < nums.size(); i++){
            maxEnding = max(nums[i], maxEnding + nums[i]);

            res = max(res, maxEnding);
        }

        return res;
    }
};

